GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m

```diff
@@ -323,17 +350,128 @@ - (void)removeEvents:(NSSet<NSNumber *> *)eventIDs {
+                                                 error:&error];
+        if (error) {
+          GDTCORLogDebug(@"An event file wasn't not moveable into the batch directory: %@", error);
+        }
+        [events addObject:event];
+      }
+    }
+    if (onComplete) {
+      onComplete(batchID, events);
+    }
+  });
+  return batchID;
+}
+
+- (void)removeBatchWithID:(nonnull NSNumber *)batchID
+             deleteEvents:(BOOL)deleteEvents
+               onComplete:(void (^_Nullable)(void))onComplete {
+  dispatch_async(_storageQueue, ^{
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    NSError *error;
+    NSArray<NSString *> *batches =
+        [fileManager contentsOfDirectoryAtPath:[GDTCORFlatFileStorage batchDataStoragePath]
+                                         error:&error];
+    if (error) {
+      if (onComplete) {
+        onComplete();
+      }
+      return;
+    }
+    for (NSString *path in batches) {
+      NSArray<NSString *> *components = [[path lastPathComponent] componentsSeparatedByString:@"."];
+      if (components.count != 2) {
+        GDTCORLogDebug(@"The name of the batch folder couldn't be parsed: %@", path);
+      }
+      NSString *targetString = components[0];
+      NSString *batchID = components[1];
+      if ([[path lastPathComponent] hasSuffix:batchID]) {
+        if (deleteEvents) {
+          NSString *deletionPath =
+              [[GDTCORFlatFileStorage batchDataStoragePath] stringByAppendingPathComponent:path];
+          [fileManager removeItemAtPath:deletionPath error:nil];
+        } else {
+          NSString *destinationPath = [[GDTCORFlatFileStorage eventDataStoragePath]
+              stringByAppendingPathComponent:targetString];
+          [fileManager moveItemAtPath:path toPath:destinationPath error:&error];
+          if (error) {
+            GDTCORLogDebug(@"Error encountered whilst moving events back: %@", error);
+          }
+        }
+      }
+    }
+    if (onComplete) {
+      onComplete();
+    }
+  });
+}
 
 - (void)libraryDataForKey:(nonnull NSString *)key
                onComplete:
-                   (nonnull void (^)(NSData *_Nullable, NSError *_Nullable error))onComplete {
+                   (nullable NSData * (^)(NSData *_Nullable, NSError *_Nullable error))onComplete {
   dispatch_async(_storageQueue, ^{
     NSString *dataPath = [[[self class] libraryDataStoragePath] stringByAppendingPathComponent:key];
     NSError *error;
     NSData *data = [NSData dataWithContentsOfFile:dataPath options:0 error:&error];
     if (onComplete) {
-      onComplete(data, error);
+      NSData *newValue = onComplete(data, error);
+      // The -isKindOfClass check is necessary because without an explicit 'return nil' in the block
+      // the implicit return value will be the block itself. The compiler doesn't detect this.
+      if (newValue && [newValue isKindOfClass:[NSData class]] && newValue.length) {
```

Comment:
What if it's not an NSData?

Reply:
I'm not sure what you mean. If it's not an NSData then we do nothing. It shouldn't be an error since it's allowable from the compiler's perspective.

Reply:
Isn't it controllable what gets returned from `onComplete`?   What else could be returned besides an `NSData`?

Reply:
The completion handler has a specific prototype (returning `nullable NSData *`), but the compiler isn't enforcing it.